From d132958a5d2ac8822946365416aea4ddb89703b5 Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Tue, 17 Jun 2025 19:53:18 +0800
Subject: [PATCH 466/960] downstream: iommu/arm-smmu-v3: Reducing the CMD_SYNC
 times

Benifiting from the hardware same address blocking mechanism.
We only need to send the CMD_SYNC when creating pte entry other
than lvl 3, this can reduce the amounts of CMD_SYNC times sent
to avoiding the errata. Detecting the hardware ability by read
the SMMU_USER_CONFIG1.

Fixes: 46f0c5798ba5 ("iommu/arm-smmu-v3: Enable iotlb_sync_map according to SMMU_IIDR")
Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
Signed-off-by: Qi Xi <xiqi2@huawei.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 14 +++++++++-----
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h |  3 +++
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index f894de2cacb4..a2de0626a9c8 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -2233,6 +2233,14 @@ static void arm_smmu_tlb_inv_page_nosync(struct iommu_iotlb_gather *gather,
 static void arm_smmu_tlb_inv_walk(unsigned long iova, size_t size,
 				  size_t granule, void *cookie)
 {
+#ifdef CONFIG_HISILICON_ERRATUM_162100602
+	struct arm_smmu_domain *smmu_domain = cookie;
+
+	if (!size && smmu_domain->smmu->options & ARM_SMMU_OPT_SYNC_BATCH) {
+		arm_smmu_tlb_inv_range_domain(iova, granule, granule, true, cookie);
+		return;
+	}
+#endif
 	arm_smmu_tlb_inv_range_domain(iova, size, granule, false, cookie);
 }
 
@@ -4350,11 +4358,7 @@ static int arm_smmu_device_hw_probe(struct arm_smmu_device *smmu)
 	}
 
 #ifdef CONFIG_HISILICON_ERRATUM_162100602
-	/* IIDR */
-	reg = readl_relaxed(smmu->base + ARM_SMMU_IIDR);
-	if (FIELD_GET(IIDR_VARIANT, reg) == 0x3 &&
-	    FIELD_GET(IIDR_REVISION, reg) == 0x2)
-		smmu->options |= ARM_SMMU_OPT_SYNC_MAP;
+	hisi_smmu_check_errata(smmu);
 #endif
 
 	if (arm_smmu_ops.pgsize_bitmap == -1UL)
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
index abf7abda5687..537f9feef2b3 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
@@ -133,6 +133,8 @@
 #define ARM_SMMU_GERROR_IRQ_CFG1	0x70
 #define ARM_SMMU_GERROR_IRQ_CFG2	0x74
 
+#define ARM_SMMU_USER_CFG1		0xe04
+
 #define ARM_SMMU_STRTAB_BASE		0x80
 #define STRTAB_BASE_RA			(1UL << 62)
 #define STRTAB_BASE_ADDR_MASK		GENMASK_ULL(51, 6)
@@ -708,6 +710,7 @@ struct arm_smmu_device {
 #define ARM_SMMU_OPT_MSIPOLL		(1 << 2)
 #define ARM_SMMU_OPT_CMDQ_FORCE_SYNC	(1 << 3)
 #define ARM_SMMU_OPT_SYNC_MAP		(1 << 4)
+#define ARM_SMMU_OPT_SYNC_BATCH		(1 << 5)
 	u32				options;
 
 	struct arm_smmu_cmdq		cmdq;
-- 
2.43.0

